bbq/compiler/peephole_pass.go

```diff
@@ -0,0 +1,182 @@
+	compiler         *Compiler[opcode.Instruction, interpreter.StaticType]
+	// bytecodeShifts is a list of where bytecode has been changed
+	// and how much the offset has changed. this is used to patch jumps
+	// after the peephole optimizations have been applied
+	bytecodeShifts []BytecodeShiftPair
+
+	// jumps is a list of offsets in the optimized bytecode that need to be patched
+	jumps []int
+}
+
+func NewPeepholeOptimizer[E, T any](compiler *Compiler[E, T]) *PeepholeOptimizer[E, T] {
+	// this map is meant to improve pattern matching performance
+	patternsByOpcode := make(map[opcode.Opcode][]PeepholePattern)
+	for _, pattern := range AllPatterns {
+		if patternsByOpcode[pattern.Opcodes[0]] == nil {
+			patternsByOpcode[pattern.Opcodes[0]] = make([]PeepholePattern, 0)
+		}
+		patternsByOpcode[pattern.Opcodes[0]] = append(patternsByOpcode[pattern.Opcodes[0]], pattern)
+	}
+	// restrict the compiler to opcode.Instruction and static type
+	// so patterns can ignore generics
+	if c, ok := any(compiler).(*Compiler[opcode.Instruction, interpreter.StaticType]); ok {
+		return &PeepholeOptimizer[E, T]{
+			PatternsByOpcode: patternsByOpcode,
+			compiler:         c,
+		}
+	}
+	return nil
+}
+
+// logic for patching jumps
+// jumps are in order of their target offset (NOT their position in the optimized bytecode),
+// bytecode shifts are also in order of their offset,
+// so we can keep a counter for cumulative shift and patch jumps in the target order
+// ASSUMPTION: peephole patterns should never match a jump instruction
+func (o *PeepholeOptimizer[E, T]) patchJumps(optimized []opcode.Instruction) {
+	cumShift := 0
+	currentShiftIndex := 0
+	for _, jump := range o.jumps {
+		var jumpTarget int
+		switch ins := optimized[jump].(type) {
+		case opcode.InstructionJump:
+			jumpTarget = int(ins.Target)
+		case opcode.InstructionJumpIfFalse:
+			jumpTarget = int(ins.Target)
+		case opcode.InstructionJumpIfTrue:
+			jumpTarget = int(ins.Target)
+		case opcode.InstructionJumpIfNil:
+			jumpTarget = int(ins.Target)
+		}
+
+		// apply all shifts up to the current jump target
+		for currentShiftIndex < len(o.bytecodeShifts) && o.bytecodeShifts[currentShiftIndex].Offset < jumpTarget {
+			cumShift += o.bytecodeShifts[currentShiftIndex].Shift
+			currentShiftIndex++
+		}
+
+		// patch jump
+		newJumpTarget := jumpTarget + cumShift
+		switch optimized[jump].(type) {
+		case opcode.InstructionJump:
+			optimized[jump] = opcode.InstructionJump{Target: uint16(newJumpTarget)}
+		case opcode.InstructionJumpIfFalse:
+			optimized[jump] = opcode.InstructionJumpIfFalse{Target: uint16(newJumpTarget)}
+		case opcode.InstructionJumpIfTrue:
+			optimized[jump] = opcode.InstructionJumpIfTrue{Target: uint16(newJumpTarget)}
+		case opcode.InstructionJumpIfNil:
+			optimized[jump] = opcode.InstructionJumpIfNil{Target: uint16(newJumpTarget)}
+		}
+	}
+}
+
+func (o *PeepholeOptimizer[E, T]) OptimizeInstructions(instructions []opcode.Instruction) []opcode.Instruction {
```

Comment:
Can you also do a quick benchmark comparison for the compilation times with and without the optimizations? Optimizations adding some compilation overhead is inevitable I guess, but we can chose to only optimize contracts (which are compiled only once and executed multiple times). A simple benchmark might help us getting some rough idea.